src/driver/pixart/pixart_pmw.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * SPDX-FileCopyrightText: 2021 Rafael Silva <silvagracarafael@gmail.com>
+ */
+
+#include "driver/pixart/pixart_pmw.h"
+
+struct motion_burst_t {
+	u8 motion;
+	u8 observation;
+	s16 dx;
+	s16 dy;
+} __attribute__((packed));
+
+u8 pixart_pmw_read(struct pixart_pmw_driver_t driver, uint8_t address)
+{
+	driver.spi_hal.select(driver.spi_hal, 1);
+
+	driver.spi_hal.transfer(driver.spi_hal, address & 0x7F); /* 7 bit address + read bit(0) */
+
+	driver.ticks_hal.delay_us(120); /* Tsrad */
+
+	u8 value = driver.spi_hal.transfer(driver.spi_hal, 0x00);
+
+	driver.spi_hal.select(driver.spi_hal, 0);
+
+	driver.ticks_hal.delay_us(160);
+
+	return value;
+}
+
+void pixart_pmw_write(struct pixart_pmw_driver_t driver, uint8_t address, uint8_t value)
+{
+	driver.spi_hal.select(driver.spi_hal, 1);
+
+	driver.spi_hal.transfer(driver.spi_hal, address | 0x80); /* 7 bit address + write bit(1) */
+
+	driver.ticks_hal.delay_us(120); /* Tsrad */
+
+	driver.spi_hal.transfer(driver.spi_hal, value);
+
+	driver.spi_hal.select(driver.spi_hal, 0);
+
+	driver.ticks_hal.delay_us(160);
+}
+
+struct motion_burst_t pixart_pmw_read_motion_burst(struct pixart_pmw_driver_t driver)
+{
+	struct motion_burst_t motion_burst;
+
+	driver.spi_hal.select(driver.spi_hal, 1);
+
+	driver.spi_hal.transfer(driver.spi_hal, PIXART_PMW_REG_BURST); /* 7 bit address + read bit(0) */
+
+	driver.ticks_hal.delay_us(35); /* Tsrad_motbr */
+
+	for (u8 i = 0; i < sizeof(motion_burst); i++)
+		((u8 *) &motion_burst)[i] = driver.spi_hal.transfer(driver.spi_hal, 0x00);
+
+	driver.spi_hal.select(driver.spi_hal, 0);
+
+	return motion_burst;
+}
+
+struct pixart_pmw_driver_t pixart_pmw_init(u8 *firmware, struct spi_hal_t spi_hal, struct ticks_hal_t ticks_hal)
+{
+	struct pixart_pmw_driver_t driver = {};
+
+	driver.spi_hal = spi_hal;
+	driver.ticks_hal = ticks_hal;
+
+	if (!firmware)
+		/*return NULL;*/ return driver;
+
+	driver.spi_hal.select(driver.spi_hal, 0);
+	driver.ticks_hal.delay_ms(1);
+
+	/* Software Reset */
+	pixart_pmw_write(driver, PIXART_PMW_REG_PWR_UP_RST, PIXART_PMW_RESET_CMD);
+	driver.ticks_hal.delay_ms(50);
+
+	/* Check Known ID */
+	u8 pid = pixart_pmw_read(driver, PIXART_PMW_REG_PID);
+	if (!((pid == PIXART_PMW_PID_PMW3360) || (pid == PIXART_PMW_PID_PMW3330) || (pid == PIXART_PMW_PID_PMW3389)))
```

Comment:
nitpick: I generally prefer the expanded De Morgan's law version (`pid != PIXART_PMW_PID_PMW3330 && pid != PIXART_PMW_PID_PMW3360 ...`)

Reply:
Works for me